README.md
# iveeCrest
a PHP library for EVE Online's CREST API (beta)

Copyright © 2015 by Aineko Macx.
All rights reserved.


## License
Unless otherwise noted, all files in this distribution are released under the LGPL v3.
See the file LICENSE included with the distribution.


## Purpose and target audience
The goal of this project is to provide its users with a simple but powerful library to access EVE's CREST API. By hiding the complexities and some of the quirks of CREST, iveeCrest helps developers to quickly prototype scripts or develop full blown (web) applications. iveeCrest will likely be most useful for developers with at least basic PHP knowledge.


## Features
- Object oriented design with high power-to-weight ratio
- Authenticated CREST based
- Methods for all endpoints reachable from CREST root are available, plus a few deeper endpoints. Easily expanded to support more (pull requests welcome!)
- Gathering of multipage responses
- Supports parallel GET requests with high performance asynchronous processing
- Multilayer cache design, supporting Memcached or Redis as caching backend
- The index-less collections returned by CREST are properly re-indexed by IDs
- Extensible via configurable subclassing
- A well documented and mostly PSR compliant codebase
- Includes a self-contained web-script to retrieve a refresh token


## Requirements
- PHP >= 5.4 (64 bit)
- php5-curl
- One of the following: Memcached + php5-memcached OR Redis + php5-redis ([phpredis](https://github.com/phpredis/phpredis))
- Developed and used under Ubuntu 14.04, other platforms might work but are untested

The necessary amount of cache memory to avoid hot data being evicted is strongly dependant on the usage pattern. Users should carefully monitor their caches (and review their applications data requirements) before hammering the CREST API with duplicate queries due to responses lost to memory pressure.


## Installation

You'll probably want to git clone iveeCrest directly into your project:

```
cd /path/to/my/project
git clone git://github.com/aineko-m/iveeCrest.git
```

### Getting a refresh token

To pull data from authenticated CREST, we'll need to acquire a refresh token, which is tied to your application and the character that authorized it's data to be used.

iveeCrest comes with a self-contained web-script that does just this. It can be found under www/getrefreshtoken.php.
Simply copy that file to a webserver and point your web browser to it. The script will take you through the steps. You'll be asked to register your application at [https://developers.eveonline.com/applications](https://developers.eveonline.com/applications) if you haven't already, selecting "CREST Access" and the "publicData" scope.

In the script form you'll then need to enter application ID, secret and redirect URL (filled automatically). When you submit the form you'll be redirected to the CREST authentication page where you'll need to authorize your app to access your characters data. With that done your browser will be redirected back to the script and it'll be able to pull the refresh token and display it. Take note, we'll need it.


### Setup iveeCrest

Make a copy of the file iveeCrest/Config_template.php, naming it Config.php and edit the configuration to match your environment.
This includes the CREST client details (CREST Base URL, client ID, client secret, refresh token), the http user agent and the cache configuration.
To switch between Memcached and Redis, apart from changing the port, you'll need to change the configured cache class. In the static array $classes find the key "Cache" and change it's value to "\iveeCrest\MemcachedWrapper" or "\iveeCrest\RedisWrapper".

That's about it. You can test the basic functionality by running the test/IveeCrestTest.php script with phpunit, which may take roughly 1 minute to run starting from a cold cache.


## Usage

Basic usage is shown with some examples:

```php
<?php
//initialize iveeCrest. Adapt path as required.
require_once('/path/to/iveeCrest/iveeCrestInit.php');

//instantiate the CREST client, passing the configured options
$client = new iveeCrest\Client(
    iveeCrest\Config::getCrestBaseUrl(),
    iveeCrest\Config::getClientId(),
    iveeCrest\Config::getClientSecret(),
    iveeCrest\Config::getUserAgent(),
    iveeCrest\Config::getClientRefreshToken()
);
//instantiate an endpoint handler
$handler = new iveeCrest\EndpointHandler($client);

//show response data from verifyAccessToken call
print_r($handler->verifyAccessToken());

//get regions endpoint
print_r($handler->getRegions());

//get specific region endpoint
print_r($handler->getRegion(10000002));

//gather all item groups (multipage response is gathered automatically)
print_r($handler->getItemGroups());

//get all market orders for Tritanium in The Forge
print_r($handler->getMarketOrders(34, 10000002));
```

These examples showcase some of the implemented endpoints in the EndpointHandler class. It is far from covering all endpoints, but more will be added over time. However, you are not forced to use it at all, you can also call the Client class directly and implement your own endpoint handling if it suits your use case better.

The Client class implements the infrastructure for getting data from CREST. This includes methods for authentication handling, simple requests to single endpoints, automatic gathering of multipage responses and parallel GET requests. The latter is a bit more complex to implement as callback functions need to be used to asynchronously process responses, but provides great flexibility and performance. EndpointHandler has a few examples of how to do it.

To help with the understanding of the structuring of the library, a [class diagram](https://github.com/aineko-m/iveeCrest/raw/master/doc/iveeCrest_class_diagram.pdf) is provided.


## Extending iveeCrest
If you extend the library with features that are generally useful and compatible with the goals and structuring of the project, Github pull requests are welcome. In any case, if you modify iveeCrest source code, you'll need to comply with the LGPL and release your modifications under the same license.

To adapt iveeCrest to your needs without changing the code, the suggested way of doing so is to use subclassing, creating new classes inheriting from the iveeCrest classes, and adapting the configuration (iveeCrest/Config::classes) accordingly. Class names are looked up dynamically, so with the adjustment objects from your classes will get instantiated instead.


## Developer notes
Although CREST is aimed at replacing not only the XML API but also the SDE, from the consumers perspective it has some weaknesses.
- Performance: Even with consumer side caching there is a general performance issue which makes live data pulling (as in "user waiting for the app") impractical except in [tailored cases](https://forums.eveonline.com/default.aspx?g=posts&t=402562). Due to the RESTful way the endpoints are structured, often dozens or hundreds of API calls are necessary to fetch just a single bit of wanted data.
- Inflexibility: CREST doesn't offer ways to filter, search or combine data like you can in a relational database, you generally have to pull in all the data in search yourself. This compounds on the performance issue.
- Availability: As CREST is tied to the live EVE cluster, it also follows the daily downtime. 3rd party API access is also among the first to be shut down in case of problems.

One of the possible solutions for these issues (independent of this library) is to collect and persist the data from CREST in a database and serve the application from there instead of pulling it live. That adds complexity to the application code but that's a price to pay for gaining resilience and ergonomics.


## Future plans
Both CREST and this library are unfinished, so significant changes should be expected with time. More endpoint handling methods will come. At some point CREST will introduce changes that will certainly make some reworking of the library necessary. More authorization scopes, for instance. iveeCrest will also be integrated into [iveeCore](https://github.com/aineko-m/iveeCore), but I intend to maintain this "solo" library in parallel.

If you find bugs, have any suggestions or other feedback or are "just" a user, please post in [this thread](https://forums.eveonline.com/default.aspx?g=posts&t=409103).


## Acknowledgements
EVE Online is a registered trademark of CCP hf.

RELEASENOTES
This file lists major changes in iveeCrest releases.

Version 0.1 - February 26th, 2015
---------------------------------
- First public beta release

test/IveeCrestTest.php
<?php
/**
 * iveeCrest PHPUnit testfile
 *
 * PHP version 5.4
 *
 * @category IveeCrest
 * @package  IveeCrestTests
 * @author   Aineko Macx <ai@sknop.net>
 * @license  https://github.com/aineko-m/iveeCrest/blob/master/LICENSE GNU Lesser General Public License
 * @link     https://github.com/aineko-m/iveeCrest/blob/master/test/IveeCrestTest.php
 *
 */

error_reporting(E_ALL);
ini_set('display_errors', 'on');

//include the iveeCrest init, expected in the iveeCrest directory, with absolute path
require_once(dirname(__FILE__) . DIRECTORY_SEPARATOR . '..' . DIRECTORY_SEPARATOR . 'iveeCrestInit.php');

/**
 * PHPUnit test for iveeCrest
 *
 * The tests some parts of iveeCrest. It is mainly used to aid during the development, but can also be used to check the
 * correct working of an iveeCrest installation. Currently the tests just call methods and check for returned type.
 *
 * To run this test, you'll need to have PHPUnit installed as well as created the iveeCrest/Config.php file based on the
 * provided template. Expect the test to run for about a minute starting with a cold cache.
 *
 * @category IveeCore
 * @package  IveeCoreExtensions
 * @author   Aineko Macx <ai@sknop.net>
 * @license  https://github.com/aineko-m/iveeCrest/blob/master/LICENSE GNU Lesser General Public License
 * @link     https://github.com/aineko-m/iveeCrest/blob/master/test/IveeCoreTest.php
 */
class IveeCrestTest extends PHPUnit_Framework_TestCase
{
    /**
     * @var \iveeCrest\Client $client
     */
    protected $client;

    /**
     * @var \iveeCrest\EndpointHandler $handler
     */
    protected $handler;

    protected function setUp()
    {
        $this->client = new iveeCrest\Client(
            iveeCrest\Config::getCrestBaseUrl(),
            iveeCrest\Config::getClientId(),
            iveeCrest\Config::getClientSecret(),
            iveeCrest\Config::getUserAgent(),
            iveeCrest\Config::getClientRefreshToken()
        );

        $this->handler = new iveeCrest\EndpointHandler($this->client);
    }

    public function testClient()
    {
        $this->assertTrue($this->client->getRootEndpoint() instanceof stdClass);
        $this->assertTrue($this->client->getCache() instanceof iveeCrest\ICache);
        $this->assertTrue($this->client->getOptions($this->client->getRootEndpointUrl()) instanceof stdClass);
    }

    public function testHandler()
    {
        $this->assertTrue($this->handler->verifyAccessToken() instanceof stdClass);
        $this->assertTrue($this->handler->tokenDecode() instanceof stdClass);
        $this->assertTrue(is_array($this->handler->getMarketTypes()));
        $this->assertTrue(is_array($this->handler->getMarketTypeHrefs()));
        $this->assertTrue(is_array($this->handler->getRegions()));
        $this->assertTrue($this->handler->getRegion(10000002) instanceof stdClass);
        $this->assertTrue(is_array($this->handler->getConstellationHrefs()));
        $this->assertTrue($this->handler->getConstellation(21000316) instanceof stdClass);
        $this->assertTrue(is_array($this->handler->getSolarSystemHrefs()));
        $this->assertTrue($this->handler->getSolarSystem(31000054) instanceof stdClass);
        $this->assertTrue($this->handler->getMarketOrders(34, 10000002) instanceof stdClass);
        $this->assertTrue(is_array($this->handler->getMarketHistory(34, 10000002)));
        $this->assertTrue(is_array($this->handler->getIndustrySystems()));
        $this->assertTrue(is_array($this->handler->getMarketPrices()));
        $this->assertTrue(is_array($this->handler->getIndustryFacilities()));
        $this->assertTrue(is_array($this->handler->getItemGroups()));
        $this->assertTrue($this->handler->getItemGroup(40) instanceof stdClass);
        $this->assertTrue(is_array($this->handler->getAlliances()));
        $this->assertTrue($this->handler->getAlliance(99000652) instanceof stdClass);
        $this->assertTrue(is_array($this->handler->getItemTypes()));
        $this->assertTrue($this->handler->getType(35) instanceof stdClass);
        $this->assertTrue(is_array($this->handler->getItemCategories()));
        $this->assertTrue($this->handler->getItemCategory(5) instanceof stdClass);
        $this->assertTrue(is_array($this->handler->getMarketGroups()));
        $this->assertTrue($this->handler->getMarketGroup(2) instanceof stdClass);
        $this->assertTrue(is_array($this->handler->getMarketGroupTypes(2)));
        $this->assertTrue(is_array($this->handler->getTournaments()));
        $this->assertTrue($this->handler->getWar(1) instanceof stdClass);
        $this->assertTrue($this->handler->getKillmail(
            'http://public-crest.eveonline.com/killmails/30290604/787fb3714062f1700560d4a83ce32c67640b1797/'
            ) instanceof stdClass
        );
    }
}


iveeCrestInit.php
<?php
/**
 * This is the single include file required for using iveeCrest in applications or scripts. All other required classes
 * are loaded via autoloader.
 *
 * PHP version 5.4
 *
 * @category IveeCrest
 * @package  IveeCrestInit
 * @author   Aineko Macx <ai@sknop.net>
 * @license  https://github.com/aineko-m/iveeCrest/blob/master/LICENSE GNU Lesser General Public License
 * @link     https://github.com/aineko-m/iveeCrest/blob/master/iveeCrest/iveeCrestInit.php
 *
 */

//Check PHP version
if (version_compare(PHP_VERSION, '5.4') < 0)
    exit('PHP Version 5.4 or higher required. Currently ' . PHP_VERSION . PHP_EOL);

//Check for 64 bit PHP
//Integers used in iveeCrest can easily exceed the maximum 2^31 of 32 bit PHP
if (PHP_INT_SIZE < 8)
    exit('64 bit PHP required. Currently ' . PHP_INT_SIZE * 8 . ' bit' . PHP_EOL);

//eve runs on UTC time
date_default_timezone_set('UTC');

//register iveeCrest's class loader
spl_autoload_register('iveeClassLoader');

/**
 * Auto class loader. Improved from http://www.php-fig.org/psr/psr-0/
 *
 * @param string $className the fully qualified class name. The loader relies on PSR compliant namespacing and class
 * file directory structuring to find and load the required files.
 *
 * @return void
 */
function iveeClassLoader($className)
{
    $className = ltrim($className, '\\');
    $fileName  = dirname(__FILE__) . DIRECTORY_SEPARATOR;
    $namespace = '';
    if ($lastNsPos = strrpos($className, '\\')) {
        $namespace = substr($className, 0, $lastNsPos);
        $className = substr($className, $lastNsPos + 1);
        $fileName  .= str_replace('\\', DIRECTORY_SEPARATOR, $namespace) . DIRECTORY_SEPARATOR;
    }
    $fileName .= str_replace('_', DIRECTORY_SEPARATOR, $className) . '.php';
    if(file_exists($fileName))
        require_once $fileName;
}

iveeCrest/Client.php
<?php
/**
 * Client class file.
 *
 * PHP version 5.4
 *
 * @category IveeCrest
 * @package  IveeCrestClasses
 * @author   Aineko Macx <ai@sknop.net>
 * @license  https://github.com/aineko-m/iveeCrest/blob/master/LICENSE GNU Lesser General Public License
 * @link     https://github.com/aineko-m/iveeCrest/blob/master/iveeCrest/Client.php
 */

namespace iveeCrest;

/**
 * The Client class provides the infrastructure for requesting data from CREST. Apart from handling authentication, it
 * offers methods for gathering and reindexing multipage endpoints as well as parallel GET with asynchronous response
 * processing.
 *
 * @category IveeCrest
 * @package  IveeCrestClasses
 * @author   Aineko Macx <ai@sknop.net>
 * @license  https://github.com/aineko-m/iveeCrest/blob/master/LICENSE GNU Lesser General Public License
 * @link     https://github.com/aineko-m/iveeCrest/blob/master/iveeCrest/Client.php
 */
class Client
{
    //the root endpoint representation
    const ROOT_REPRESENTATION = 'vnd.ccp.eve.Api-v3+json';

    /**
     * @var \iveeCrest\ICache $cache for data objects
     */
    protected $cache;

    /**
     * @var string $rootEndpointUrl specifies the CREST root URL from which most other endpoints can be navigated to.
     */
    protected $rootEndpointUrl;

    /**
     * @var string $clientId holds the client application ID as configured in CCPs developer application backend:
     * https://developers.eveonline.com/applications
     */
    protected $clientId;

    /**
     * @var string $clientSecret the secret key to go along with the client application ID.
     */
    protected $clientSecret;

    /**
     * @var string $charRefreshToken this is the character specific and durable refresh token.
     */
    protected $charRefreshToken;

    /**
     * @var string $charAccessToken this is the character specific and short lived access token, gotten by using the
     * refresh token.
     */
    protected $charAccessToken;

    /**
     * @var int $charAccessTokenExpiry timestamp when the access token expires (and will need to be refreshed).
     */
    protected $charAccessTokenExpiry;

    /**
     * @var \iveeCrest\CurlWrapper $cw holds the object handling CURL.
     */
    protected $cw;

    /**
     * @var \stdClass $rootEndpoint holds the root endpoint (after having been requested at least once)
     */
    protected $rootEndpoint;

    /**
     * Constructs a Client object. Note that these are character-specific.
     *
     * @param string $rootEndpointUrl the Url to the CREST root
     * @param string $clientId the Id of the app you registered
     * @param string $clientSecret the secret for the app you registered
     * @param string $clientUserAgent the user agent that should be used in the requests
     * @param string $charRefreshToken the chracter-specific refresh token to be used
     *
     * @return \iveeCrest\Client
     */
    public function __construct($rootEndpointUrl, $clientId, $clientSecret, $clientUserAgent, $charRefreshToken)
    {
        $this->rootEndpointUrl    = $rootEndpointUrl;
        $this->clientId           = $clientId;
        $this->clientSecret       = $clientSecret;
        $this->charRefreshToken   = $charRefreshToken;

        $cacheClass = Config::getIveeClassName('Cache');
        $this->cache = new $cacheClass(get_called_class() . '_' . $charRefreshToken);

        $cwClass = Config::getIveeClassName('CurlWrapper');
        $this->cw = new $cwClass($clientUserAgent, $charRefreshToken);

        //load root endpoint
        $this->rootEndpoint = $this->getEndpoint(
            $this->rootEndpointUrl,
            false,
            static::ROOT_REPRESENTATION
        );
    }

    /**
     * Returns the root endpoint.
     *
     * @return \stdClass
     */
    public function getRootEndpointUrl()
    {
        return $this->rootEndpointUrl;
    }

    /**
     * Returns the used cache object.
     *
     * @return \iveeCrest\ICache
     */
    public function getCache()
    {
        return $this->cache;
    }

    /**
     * Returns a basic authorization header.
     *
     * @return array
     */
    protected function getBasicAuthHeader()
    {
        return array('Authorization: Basic ' . base64_encode($this->clientId . ':' . $this->clientSecret));
    }

    /**
     * Returns a bearer authorization header, pulling a new access token if necessary.
     *
     * @return array
     */
    protected function getBearerAuthHeader()
    {
        return array('Authorization: Bearer ' . $this->getAccessToken());
    }

    /**
     * Returns the necessary POST fields to request a new access token.
     *
     * @return array
     */
    protected function getRefreshTokenPostFields()
    {
        return array(
            'grant_type' => 'refresh_token',
            'refresh_token' => $this->charRefreshToken
        );
    }

    /**
     * Returns an access token, requesting a new one if none available or expired.
     *
     * @return string
     */
    protected function getAccessToken()
    {
        //if we don't have an access token, get one
        if (!isset($this->charAccessToken) OR time() >= $this->charAccessTokenExpiry) {
            $accessTokenResponse = $this->cw->post(
                $this->getRootEndpoint()->authEndpoint->href, 
                $this->getBasicAuthHeader(), 
                $this->getRefreshTokenPostFields()
            );
            $this->charAccessToken = $accessTokenResponse->content->access_token;
            $this->charAccessTokenExpiry = time() + $accessTokenResponse->content->expires_in - 10;
        }
        return $this->charAccessToken;
    }

    /**
     * Returns the root endpoint.
     *
     * @return \stdClass
     */
    public function getRootEndpoint()
    {
        return $this->rootEndpoint;
    }

    /**
     * Returns the data returned by an OPTIONS call to a CREST endpoint.
     *
     * @param string $url to be used
     *
     * @return \stdClass
     */
    public function getOptions($url)
    {
        return $this->cw->options($url)->content;
    }

    /**
     * Performs a GET request to a CREST endpoint, returning the full response object.
     *
     * @param string $url the URL of the endpoint
     * @param bool $auth if authentication header should be sent
     * @param string $accept the requested representation
     *
     * @return \iveeCrest\Response
     */
    public function getEndpointResponse($url, $auth = false, $accept = null)
    {
        if($auth)
            $header = $this->getBearerAuthHeader();
        else
            $header = array();

        if(isset($accept))
            $header[] = 'Accept: application/' . $accept;

        return $this->cw->get($url, $header);
    }

    /**
     * Performs a GET request to a CREST endpoint, returning data from the response.
     *
     * @param string $url the URL of the endpoint
     * @param bool $auth if authentication header should be sent
     * @param string $accept the requested representation
     *
     * @return \stdClass
     */
    public function getEndpoint($url, $auth = false, $accept = null)
    {
        return $this->getEndpointResponse($url, $auth, $accept)->content;
    }

    /**
     * Gathers multipage endpoint responses and joins them into one array, using the passed callback functions to 
     * traverse and index the data. Since this operation is potentially expensive, it is recommended to use
     * gatherCached() instead, which introduces another layer of caching.
     *
     * @param string $endpointHref the URL to the first page of the endpoint 
     * @param callable $indexFunc function to be used to extract the ID from/for and individual response item
     * @param callable $elementFunc function to be used to extract the desired data from and individual response item
     * @param string $accept the representation to request from CREST
     *
     * @return array
     */
    public function gather($endpointHref, callable $indexFunc = null, callable $elementFunc = null, $accept = null)
    {
        $ret = array();
        $href = $endpointHref;

        while (true) {
            //get the response for the current href
            $response = $this->getEndpointResponse($href, true, $accept);

            foreach ($response->content->items as $item) {
                //if an element function has been given, call it, otherwise use the full item in the result array
                if(is_null($elementFunc))
                    $element = $item;
                else
                    $element = $elementFunc($item);

                //if an index function has been given, call it to get a result key, otherwise just push the element
                //onto result array
                if(is_null($indexFunc))
                    $ret[] = $element;
                else
                    $ret[$indexFunc($item)] = $element;
            }

            //if there are more pages, do another iteration with updated href
            if ($response->hasNextPage())
                $href = $response->getNextPageHref();
            else
                break;
        }
        return $ret;
    }

    /**
     * Gathers multipage endpoint responses and joins them into one array, using the passed callback functions to 
     * traverse and index the data. The result of this (potentially expensive) operation is cached.
     *
     * @param string $endpointHref the URL to the first page of the endpoint 
     * @param callable $indexFunc function to be used to extract the ID from/for and individual response item
     * @param callable $elementFunc function to be used to extract the desired data from and individual response item
     * @param string $accept the representation to request from CREST
     * @param int $ttl the time to live to be used in the cache
     * @param string $subCommandKey to avoid cache namespace collisions when different gather requests access the same
     * endpoint URL, an additional subcommand key can be specified.
     *
     * @return array
     */
    public function gatherCached($endpointHref, callable $indexFunc = null, callable $elementFunc = null, 
        $accept = null, $ttl = 3600, $subCommandKey = null
    ) {
        $dataKey = 'gathered:' . $endpointHref . (isset($subCommandKey) ? ',' . $subCommandKey : '');
        //we introduce another caching layer here because gathering and reindexing multipage data is expensive, even
        //when the individual CREST responses are already cached.
        try {
            $dataObj = $this->cache->getItem($dataKey);
        } catch (Exceptions\KeyNotFoundInCacheException $e) {
            //setup a cacheable array object
            $dataClass = Config::getIveeClassName('CacheableArray');
            $dataObj = new $dataClass($dataKey, $ttl);

            //gather all the pages into one compact array
            $dataObj->data = $this->gather(
                $endpointHref,
                $indexFunc,
                $elementFunc,
                $accept
            );
            $this->cache->setItem($dataObj);
        }
        return $dataObj->data;
    }

    /**
     * Performs parallel asyncronous GET requests to a CREST endpoint. Since the same header is used for all requests,
     * all hrefs passed should be to the same endpoint. This method has void return, instead, responses are
     * passed to the callback functions provided as arguments.
     *
     * This method will most likely be most useful in batch scripting scenarios. If the same data is requested less
     * frequently than the cache TTL, it is advisable to disable caching via argument to avoid overflowing the cache
     * with data that won't be requested again before they expire.
     *
     * @param array $hrefs the hrefs to request
     * @param callable $callback a function expecting one \iveeCrest\Response object as argument, called for every
     * successful response
     * @param callable $errCallback a function expecting one \iveeCrest\Response object as argument, called for every
     * non-successful response
     * @param string $accept the requested representation
     * @param bool $cache whether the individual Responses should be cached.
     *
     * @return void
     * @throws \iveeCrest\Exceptions\IveeCrestException on general CURL error
     */
    public function asyncGetMultiEndpointResponses(array $hrefs, callable $callback, callable $errCallback = null,
        $accept = null, $cache = true
    ) {
        $header = array();
        if(isset($accept))
            $header[] = 'Accept: application/' . $accept;
        
        //can't pass "this" in callables
        $client = $this;

        //run the multi GET
        return $this->cw->asyncMultiGet(
            array_unique($hrefs),
            $header,
            function () use ($client) {
                return $client->getBearerAuthHeader(); //little trick to avoid having to make the method public
            },
            $callback,
            $errCallback,
            $cache
        );
    }
}

iveeCrest/Config_template.php
<?php
/**
 * Main configuration file for iveeCrest.
 *
 * Copy and edit this file according to your environment. The edited file should be saved as Config.php
 *
 * PHP version 5.4
 *
 * @category IveeCrest
 * @package  IveeCrestClasses
 * @author   Aineko Macx <ai@sknop.net>
 * @license  https://github.com/aineko-m/iveeCrest/blob/master/LICENSE GNU Lesser General Public License
 * @link     https://github.com/aineko-m/iveeCrest/blob/master/iveeCrest/Config_template.php
 */

namespace iveeCrest;

/**
 * The Config class holds the basic iveeCrest configuration for cache and classnames.
 *
 * @category IveeCrest
 * @package  IveeCrestClasses
 * @author   Aineko Macx <ai@sknop.net>
 * @license  https://github.com/aineko-m/iveeCrest/blob/master/LICENSE GNU Lesser General Public License
 * @link     https://github.com/aineko-m/iveeCrest/blob/master/iveeCrest/Config_template.php
 */
class Config
{
    /////////////////////
    // Edit below here //
    /////////////////////

    //Cache config
    protected static $cacheHost   = 'localhost';
    protected static $cachePort   = '11211'; //memcached default: 11211, redis default: 6379
    protected static $cachePrefix = 'iveeCrest_';

    //CREST config
    protected static $crestBaseUrl       = 'https://crest-tq.eveonline.com/';
    protected static $clientId           = 'myClientId';
    protected static $clientSecret       = 'myClientSecret';
    protected static $clientRefreshToken = 'myCharRefreshToken';

    //change the application name in the parenthesis to your application. It is used when accessing the CREST API.
    protected static $userAgent = 'iveeCrest/0.1 (unknown application)';

    //To enable developers to extend iveeCrest with their own classes (inheriting from iveeCrest), it dynamically looks
    //up class names before instantiating them. This array maps from class "nicknames" to fully qualified names, which
    //can then be used by the autoloader. Change according to your needs.
    protected static $classes = array(
        'Cache'                 => '\iveeCrest\MemcachedWrapper', //change to '\iveeCrest\RedisWrapper' if using Redis
        'CacheableArray'        => '\iveeCrest\CacheableArray',
        'Client'                => '\iveeCrest\Client',
        'CurlWrapper'           => '\iveeCrest\CurlWrapper',
        'EndpointHandler'       => '\iveeCrest\EndpointHandler',
        'Response'              => '\iveeCrest\Response',
        'CrestException'              => '\iveeCrest\Exceptions\CrestException',
        'InvalidArgumentException'    => '\iveeCrest\Exceptions\InvalidArgumentException',
        'IveeCrestException'          => '\iveeCrest\Exceptions\IveeCrestException',
        'KeyNotFoundInCacheException' => '\iveeCrest\Exceptions\KeyNotFoundInCacheException',
    );

    ////////////////////////////
    // Do not edit below here //
    ////////////////////////////

    /**
     * Instantiates Config object. Private so this class is only used as static.
     *
     * @return Config
     */
    private function __construct()
    {
    }

    /**
     * Returns if cache use is configured or not
     *
     * @return bool
     */
    public static function getUseCache()
    {
        return static::$useCache;
    }

    /**
     * Returns configured cache host name
     *
     * @return string
     */
    public static function getCacheHost()
    {
        return static::$cacheHost;
    }

    /**
     * Returns configured cache port
     *
     * @return int
     */
    public static function getCachePort()
    {
        return static::$cachePort;
    }

    /**
     * Returns configured cache prefix for keys stored by iveeCrest
     *
     * @return string
     */
    public static function getCachePrefix()
    {
        return static::$cachePrefix;
    }

    /**
     * Returns configured CREST base URL
     *
     * @return string
     */
    public static function getCrestBaseUrl()
    {
        return static::$crestBaseUrl;
    }

    /**
     * Returns configured CREST client ID
     *
     * @return string
     */
    public static function getClientId()
    {
        return static::$clientId;
    }

    /**
     * Returns configured CREST client secret
     *
     * @return string
     */
    public static function getClientSecret()
    {
        return static::$clientSecret;
    }

    /**
     * Returns configured CREST user specific refresh token
     *
     * @return string
     */
    public static function getClientRefreshToken()
    {
        return static::$clientRefreshToken;
    }

    /**
     * Returns configured user agent to be used by the CREST client
     *
     * @return string
     */
    public static function getUserAgent()
    {
        return static::$userAgent;
    }

    /**
     * Returns the fully qualified name of classes to instantiate for a given class nickname. This is used extensively
     * in iveeCrest to allow for configurable class instantiation
     *
     * @param string $classNickname a short name for the class
     *
     * @return string
     */
    public static function getIveeClassName($classNickname)
    {
        if (isset(static::$classes[$classNickname]))
            return static::$classes[$classNickname];
        else
            exit('Fatal Error: No class configured  for "' . $classNickname . '" in iveeCrest' . DIRECTORY_SEPARATOR
                . 'Config.php' . PHP_EOL);
    }
}
